### PR TITLE
Prioritize "readme.md"

### LINKED ISSUE #5678: Repository previews the wrong README.
- Gitea version (or commit ref): 1.7.0
- Git version: 2.17.1
- Operating system: 18.04.1 LTS
- Database:
  - [x] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Example provided

## Description

Repository previews the wrong README. 

Steps to reproduce:

1. Create new repo
2. Add 2 files with markdown content: `README.md` and `README.i18n.md`
3. Repository index/root will show `README.i18n.md`.

Live example observable here: [link](https://git.pwned.systems/monero/monero)

### PR DESCRIPTION
**Pull request by wewark**
Fixes #5678 
Check all the files in the home dir:
- If a file matches exactly "readme.md", prioritize it over the others,
- Otherwise, choose any file that matches "readme.md" but not strictly

### CODE REVIEW

File: routers/repo/view.go [hunk 1]
```diff
@@ -62,13 +62,13 @@ func renderDirectory(ctx *context.Context, treeLink string) {
...
-			continue
+		if markup.IsReadmeFile(entry.Name(), true) {
+			readmeFile = entry.Blob()
+			break
 		}
 
-		readmeFile = entry.Blob()
-		if markup.Type(entry.Name()) != "" {
-			break
+		if markup.IsReadmeFile(entry.Name(), false) {
```
Comment (lunny) [inline routers/repo/view.go]: We need a priority of readme file here: `readme.md` > `readme.txt` > `readme` > `readme.<i18n>.md` > ... ?

**Review by lunny — CHANGES_REQUESTED** [review]

File: routers/repo/view.go (hunk 1 above)
Comment (wewark) [reply routers/repo/view.go]: Done

File: routers/repo/view.go [hunk 2]
```diff
@@ -56,18 +56,28 @@ func renderDirectory(ctx *context.Context, treeLink string) {
 		return
 	}
 
-	var readmeFile *git.Blob
+	var readmeFiles [4]*git.Blob
```
Comment (kolaente) [inline routers/repo/view.go:62]: How can you be sure this will always be 4?

**Review by kolaente — CHANGES_REQUESTED** [review]

File: docker/usr/bin/entrypoint [hunk 3]
```diff
@@ -32,7 +32,6 @@ done
 
 if [ -f /data/gitea/conf/app.ini ]; then
     echo "Found app.ini config file, migrating database"
-    chmod 644 /data/gitea/conf/app.ini
```
[the remark this answers is not in the document]
Comment (wewark) [reply docker/usr/bin/entrypoint]: I didn't, I accidentally clicked "update branch", it's undone now.

File: options/locale/locale_fr-FR.ini [hunk 4]
```diff
@@ -561,6 +561,7 @@ migrate.lfs_mirror_unsupported=La synchronisation des objets LFS n'est pas suppo
 mirror_from=miroir de
 forked_from=bifurqué depuis
 fork_from_self=Vous ne pouvez pas bifurquer un dépôt que vous possédez.
+fork_guest_user=Connectez vous pour forker ce dépôt.
```
[the remark this answers is not in the document]
Comment (wewark) [reply options/locale/locale_fr-FR.ini]: Due to the "update branch" as well.

File: routers/repo/view.go (hunk 2 above)
Comment (wewark) [reply routers/repo/view.go:62]: This isn't for all the readme files found, the first 3 will hold files matching `{".md", ".txt", ""}` in order, the last one is for any non-strictly matching readme file. The first non-nil file in the array will be taken. This makes the priority `readme.md > readme.txt > readme > readme.<i18n>.md > ...`

File: modules/markup/markup_test.go [hunk 5]
```diff
@@ -37,4 +37,24 @@ func TestMisc_IsReadmeFile(t *testing.T) {
 	for _, testCase := range falseTestCases {
 		assert.False(t, IsReadmeFile(testCase))
 	}
+
+	trueTestCasesStrict := [][]string{
+		{"readme", ""},
+		{"readme.md", ".md"},
+		{"readme.txt", ".txt"},
+	}
+	falseTestCasesStrict := [][]string{
```
[the remark this answers is not in the document]
Comment (wewark) [reply modules/markup/markup_test.go:47]: This tests the new `IsReadmeFile()` the issue itself can't be tested here. I've added more test cases anyway.

**Review by kolaente — APPROVED** [review]

**Review by lunny — APPROVED** [review]

File: routers/repo/view.go [hunk 6]
```diff
@@ -56,18 +56,31 @@ func renderDirectory(ctx *context.Context, treeLink string) {
 		return
 	}
 
-	var readmeFile *git.Blob
+	// 3 for the extensions in exts[] in order
+	// the last one is for a readme that doesn't
+	// strictly match an extension
+	var readmeFiles [4]*git.Blob
+	var exts = []string{".md", ".txt", ""} // sorted by priority
```
Comment (techknowlogick) [inline routers/repo/view.go:63]: Can other file extensions be added: https://stackoverflow.com/a/41112364

Comment (techknowlogick) [reply routers/repo/view.go:63]: but it can output the raw content, and if a user has added an external renderer then we don't want them to miss out.

Comment (wewark) [reply routers/repo/view.go:63]: @techknowlogick All readme files are detected, it is just about the priority, there is a 4th element in `readmeFiles[]` in which any file named `readme` will be detected regardless of its extension.

Comment (techknowlogick) [reply routers/repo/view.go:63]: Sounds good.

**Pull request merged**